[mlir][vector] Add pattern to break down reductions into arith ops

kuhar opened the pull request:
The number of vector elements considered 'small' enough to extract is               
parameterized.                                                   
                                                                 
This is to avoid going into specialized reduction lowering when a
single/couple of arith ops can do. Targets without dedicated reduction  
intrinsics can use that as an emulation path too.                  
                                                                   
Depends on https://github.com/llvm/llvm-project/pull/75846.

kuhar committed:
[mlir][vector] Add pattern to break down small reductions into arith ops

The number of vector elements considered 'small' enough to extract is
parameterized.

This is to avoid going into specialized reduction lowering when a
single/couple of arith ops can do. Targets without dedicated reduction
intrinsics can use that as an emulation path too.

Depends on https://github.com/llvm/llvm-project/pull/75846.

 Please enter the commit message for your changes. Lines starting

Hardcode84 reviewed (commented):
The pattern itself is LGTM, but I'm not very familiar with this minnum/minimum buisiness.

kuhar commented:
BTW, I forgot to add to the PR description that I plan to use it to remove most of the vector reduction to SPIR-V lowering patterns, only leaving the specialized ones that we can expand to vector dot product.

banach-space commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -6212,6 +6214,7 @@ bool WarpExecuteOnLane0Op::areTypesCompatible(Type lhs, Type rhs) {
 
 Value mlir::vector::makeArithReduction(OpBuilder &b, Location loc,
                                        CombiningKind kind, Value v1, Value acc,
+                                       arith::FastMathFlagsAttr fastmath,
```
Isn't this unrelated to this change?

kuhar replied:
I need this to propagate fastmath attributes. I figured this is a relatively small change, but I can split it into a separate change if this makes code review difficult -- let me know.

banach-space replied:
The noise level is low, so that's not a concern. However, these attributes might affect things in a very subtle way and that's why I feel the deserve a dedicated commit. I would happily review that and accept (it's very straightforward).

kuhar replied:
Sure, here it is https://github.com/llvm/llvm-project/pull/75846

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp:
```diff
@@ -1578,6 +1580,50 @@ struct ReduceRedundantZero final : OpRewritePattern<vector::ReductionOp> {
+  LogicalResult matchAndRewrite(vector::ReductionOp op,
+                                PatternRewriter &rewriter) const override {
+    VectorType type = op.getSourceVectorType();
+    if (type.isScalable() || op.isMasked())
+      return failure();
```
Could you add a test for scalable vectors and masked ops? I will often rely on tests to gauge the level of support.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp:
```diff
@@ -1578,6 +1580,50 @@ struct ReduceRedundantZero final : OpRewritePattern<vector::ReductionOp> {
   }
 };
 
```
Could you document this pattern with an example?

banach-space reviewed (approved):
LGTM % some comments, thanks!

kuhar commented:
Split off the `makeArithReduction` change into https://github.com/llvm/llvm-project/pull/75727 and rebased onto it.

Added example as suggested by @banach-space.
